### PR TITLE
Get Model Inputs and Outputs: Labels and Y Predicate

### PR DESCRIPTION
**Pull request by NihalHarish**
### Description of changes:
- This PR runs only with a modified AWS TF, which supplies `x`, `y` and `y_pred` through logs.
- A new function `_save_model_outputs` that saves outputs to `OUTPUTS` collection 
- Updates to the assert statements to check for the additional tensors we save with this change.

### Performance Testing:
Machine: p3.16x
epochs=5000, steps_per_epoch=10
save_config = SaveConfig(save_interval=500)

Before:
Outputs are not saved:
Uses TF without change
```
real	20m32.546s
user	409m1.362s
sys	456m32.768s
```

With PR: 
Outputs saved:
Uses TF with change
```
real	21m0.540s
user	414m49.772s
sys	461m42.581s
```
#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: smdebug/tensorflow/base_hook.py [hunk 1]
```diff
@@ -437,7 +437,7 @@ def _write_for_tensor(self, tensor_name, tensor_value, save_collections, tensor_
 
         # this tensor_name is tf tensor name, need to convert to export_name
         tensor_ref = self._get_tensor_ref(tensor_name, save_collections=save_collections)
-        if tensor_ref:
+        if tensor_ref is not None:
```
Comment (NihalHarish) [inline smdebug/tensorflow/base_hook.py:440]: Unrelated to PR.
This if check sometimes leads to an annoying error during development.
See: https://stackoverflow.com/questions/7816363/if-a-vs-if-a-is-not-none for more details

**Comment by Vikas-kum** [comment]
Please fix the build failure.

File: smdebug/tensorflow/keras.py [hunk 2]
```diff
@@ -366,6 +366,21 @@ def _add_metric(self, metric_name, metric_value: tf.Tensor = None):
...
+        if self._is_collection_being_saved_for_step(CollectionKeys.OUTPUTS):
+            export_names = {"y_pred": "train_output/y_pred", "y": "train_output/y"}
+            self._initialize_writers(only_initialize_if_missing=True)
+            output_collection = self.collection_manager.get(CollectionKeys.OUTPUTS)
+            for key in logs:
+                if key in ["y", "y_pred"]:
+                    tensor_ref = TensorRef.from_non_graph_var(export_names[key])
+                    output_collection.set_tensor_ref(tensor_ref)
+                    self.tensor_to_collections[export_names[key]] = {output_collection}
+                    self._save_for_tensor(export_names[key], logs[key], check_before_write=False)
```
Comment (Vikas-kum) [inline smdebug/tensorflow/keras.py:382]: can we create a save_tensor function. 
save_tensor(name, value, collection=Collection.DEFAULT_tensors)

File: smdebug/tensorflow/keras.py [hunk 3]
```diff
@@ -375,7 +390,7 @@ def _save_metrics(self, batch, logs, force_save=False):
             self._initialize_writers(only_initialize_if_missing=True)
             logs["batch"] = batch
             for key in logs:
-                if key in ["loss", "val_loss", "outputs"]:
+                if key in ["loss", "val_loss", "outputs", "y", "y_pred"]:
```
Comment (Vikas-kum) [inline smdebug/tensorflow/keras.py:393]: y ypred is special name used at 374 378 and here. Can we make it as named constant?

File: smdebug/tensorflow/keras.py [hunk 4]
```diff
@@ -479,6 +495,7 @@ def on_epoch_end(self, batch, logs=None):
         if self._is_not_supported():
             return
         self._save_metrics(batch=batch, logs=logs, force_save=True)
+        self._save_model_outputs(logs=logs)
```
Comment (Vikas-kum) [inline smdebug/tensorflow/keras.py:498]: if save tensor post step has this logic, why do we need to do it here?

**Comment by Vikas-kum** [comment]
Can you comment what is the value and y and y_pred looks like ?

**Comment by NihalHarish** [comment]
> Please fix the build failure.

The build failure is due to the fact that the pipelines currently do not use the custom AWS TF version

File: smdebug/tensorflow/keras.py (hunk 4 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py]: We do not need it here, since we currently save values produced during the train step.

File: smdebug/tensorflow/keras.py (hunk 3 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py]: I have added constants

File: smdebug/tensorflow/keras.py (hunk 2 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:386]: An API for the customers to use?
We can.

**Comment by NihalHarish** [comment]
> Can you add to testing section indescription what is and y and y_pred value looks like ?
> trial.tensor("y").steps()
> trial.tensor("y_pred").value(10)

I have updated the test with additional asserts:

```
        for tname in trial.tensor_names(collection=CollectionKeys.OUTPUTS):
            output = trial.tensor(tname)
            assert tname in [ModelOutput.Y_PRED, ModelOutput.Y]
            assert output.value(0) is not None
            assert output.steps() == trial.steps(mode=ModeKeys.TRAIN)
        # Check the shape of output tensors
        assert trial.tensor(ModelOutput.Y).value(0).shape[1] == 1  # label
        assert (
            trial.tensor(ModelOutput.Y_PRED).value(0).shape[1] == 10
        )  # Output probability for each class
```

File: smdebug/tensorflow/keras.py (hunk 4 above)
Comment (Vikas-kum) [reply smdebug/tensorflow/keras.py]: Why not other stages (EVAL & predict) ?

**Comment by Vikas-kum** [comment]
> > Please fix the build failure.
> 
> The build failure is due to the fact that the pipelines currently do not use the custom AWS TF version

So, how are we taking care of this? We can't have the test failing. You need to have separate test and enable those only when we have AWS TF changes.

File: tests/tensorflow2/test_keras_mirrored.py [hunk 5]
```diff
@@ -256,8 +256,10 @@ def test_save_all(out_dir, tf_eager_mode, workers):
     tr = create_trial_fast_refresh(out_dir)
     print(tr.tensor_names())
     if tf_eager_mode:
-        assert len(tr.tensor_names()) == (6 + 2 + 1 + 5 + 1 if is_tf_2_2() else 6 + 3 + 1 + 5 + 1)
-        # weights, metrics, losses, optimizer variables, scalar
+        assert len(tr.tensor_names()) == (
+            6 + 2 + 1 + 5 + 1 + 2 if is_tf_2_2() else 6 + 3 + 1 + 5 + 1 + 2
```
Comment (Vikas-kum) [inline tests/tensorflow2/test_keras_mirrored.py:260]: this type of test 6+ 2 + 1 has no value, how do we know what is six tensors what is 2 tensors ? It seems cryptic and difficult to maintain.
We should have asserted tname.count == 6
and blah blah, rather than adding 6 + 2 + 1 ...

File: tests/tensorflow2/test_keras.py [hunk 6]
```diff
@@ -396,7 +397,8 @@ def test_gradtape_persistent(out_dir, saveall):
 @pytest.mark.slow
 @pytest.mark.parametrize("saveall", [True, False])
 def test_keras_fit(out_dir, tf_eager_mode, saveall):
-    hook = smd.KerasHook(out_dir=out_dir, save_all=saveall)
+    save_config = SaveConfig(save_interval=1) if saveall else None
```
Comment (vandanavk) [inline tests/tensorflow2/test_keras.py:400]: might be overkill for CI?

**Comment by vandanavk** [comment]
> > Please fix the build failure.
> 
> The build failure is due to the fact that the pipelines currently do not use the custom AWS TF version

so the tests would fail on vanilla TF too?

File: tests/tensorflow2/test_keras.py (hunk 6 above)
Comment (NihalHarish) [reply tests/tensorflow2/test_keras.py:400]: We run only 10 steps 

> > > Please fix the build failure.
> > 
> > 
> > The build failure is due to the fact that the pipelines currently do not use the custom AWS TF version
> 
> so the tests would fail on vanilla TF too?

Yes, this feature will only be supported on AWS TF

Comment (vandanavk) [reply tests/tensorflow2/test_keras.py:400]: so the vanilla container codebuild will fail. can you add checks to ensure that it works for both versions of TF ?

File: smdebug/core/reductions.py [hunk 7]
```diff
@@ -29,14 +29,15 @@ def get_basic_numpy_reduction(reduction_name, numpy_data):
...
 
-        if abs:
-            rv = np.linalg.norm(np.absolute(numpy_data), ord=ord)
-        else:
-            rv = np.linalg.norm(numpy_data, ord=ord)
+        if np.isscalar(numpy_data):
+            # np.linalg.norm expects array-like inputs
+            # but numpy_data can sometimes be a scalar value
+            numpy_data = [numpy_data]
+        rv = np.linalg.norm(numpy_data, ord=order)
```
Comment (vandanavk) [inline smdebug/core/reductions.py:40]: now that your PR is merged, you dont need this here right?

File: smdebug/tensorflow/keras.py [hunk 8]
```diff
@@ -366,6 +374,42 @@ def _add_metric(self, metric_name, metric_value: tf.Tensor = None):
...
+
+        for key in logs:
+            if key in [ModelOutput.Y, ModelOutput.Y_PRED]:
+                collections_to_save = self._get_collections_to_save_for_step()
+                output_collection = self.get_collection(CollectionKeys.OUTPUTS)
+                if output_collection in collections_to_save:
+                    collections_to_write = {output_collection}
+                    for collection in collections_to_save:
+                        if match_inc(export_names[key], collection.include_regex):
+                            collections_to_write.add(collection)
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:396]: isn't this done in save_for_Tensor already?

```
save_collections = self._get_collections_with_tensor(tensor_name)
        save_collections_for_tensor = save_collections.intersection(
            self._get_collections_to_save_for_step()
        )
```

File: smdebug/tensorflow/keras.py [hunk 9]
```diff
@@ -95,6 +97,11 @@ def _is_not_supported(self):
                 self._hook_supported = False
         return not self._hook_supported
 
+    def should_save_global_step_for_mode(self, mode: str):
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:100]: where is this used?

File: smdebug/core/modes.py [hunk 10]
```diff
@@ -14,3 +14,16 @@ class ModeKeys(Enum):
 ALLOWED_MODE_NAMES = [x.name for x in ALLOWED_MODES]
 MODE_STEP_PLUGIN_NAME = "mode_step"
 MODE_PLUGIN_NAME = "mode"
+
+
+def str_to_mode_keys(s):
```
Comment (vandanavk) [inline smdebug/core/modes.py:19]: this is being used only in should_save_global_step_for_mode, which doesn't have a caller. is this required?

Comment (NihalHarish) [reply smdebug/core/modes.py:19]: `should_save_global_step_for_mode` is called in aws-tf. See https://tiny.amazon.com/ljsp41dd/IsenLink for details.

File: smdebug/core/reductions.py (hunk 7 above)
Comment (NihalHarish) [reply smdebug/core/reductions.py:40]: Weird. This diff is wrong. The RHS doesn't reflect the current state of master.
